### PR TITLE
feat:  support PostgreSQL wire protocol

### LINKED ISSUE #711: Support PostgreSQL wire protocol
### Describe This Problem

MySQL wire protocol is already supported by CeresDB, it would be great to support PostgreSQL also.

### Proposal

Support PostgreSQL protocol like MySQL

### Additional Context

https://github.com/sunng87/pgwire

### PR DESCRIPTION
**Pull request by holicc**
## Rationale

close #711 

I have to create a new PR again, because the old branch name is not valid for sdk-testing.

## Detailed Changes

support PostgreSQL wire Protocol

## Test Plan

Integration tests

### CODE REVIEW

**Comment by tanruixiang** [comment]
Hi, After https://github.com/CeresDB/ceresdb/pull/1134, the apache template has changed. I updated the conflict for you so as not to confuse you.

**Comment by holicc** [comment]
@tanruixiang Hi, the CI/unit-test doesn't have enough device space. I'm not sure if the problem is with my code or the CI, but I can pass the test on my laptop. Can you help me fix this error?

`IO error: No space left on device: While appending to file: /tmp/.tmp2FtmjK/wal/000004.dbtmp.plain: No space left on device`

**Comment by tanruixiang** [comment]
> @tanruixiang Hi, the CI/unit-test doesn't have enough device space. I'm not sure if the problem is with my code or the CI, but I can pass the test on my laptop. Can you help me fix this error?
> 
> `IO error: No space left on device: While appending to file: /tmp/.tmp2FtmjK/wal/000004.dbtmp.plain: No space left on device`

Don't worry, it's not a problem with your code, I'll re-run it and try.

File: server/src/postgresql/handler.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+#[async_trait]
+impl<Q: QueryExecutor + 'static> SimpleQueryHandler for PostgresqlHandler<Q> {
+    async fn do_query<'a, C>(&self, _client: &C, sql: &'a str) -> PgWireResult<Vec<Response<'a>>>
+    where
+        C: ClientInfo + Unpin + Send + Sync,
+    {
+        let ctx = self.create_ctx().map_err(|e| {
+            PgWireError::UserError(Box::new(pgwire::error::ErrorInfo::new(
+                "ERROR".to_owned(),
+                "08000".to_owned(),
```
Comment (jiacai2050) [inline server/src/postgresql/handler.rs:50]: What does this mean? better to add some comments.

File: server/src/postgresql/handler.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        Datum::String(s) => encoder.encode_field(&s.as_str()),
+        Datum::Int16(i) => encoder.encode_field(&i),
+        Datum::Int8(i) => encoder.encode_field(&i),
+        Datum::Int32(i) => encoder.encode_field(&i),
+        Datum::Int64(i) => encoder.encode_field(&i),
+        Datum::UInt32(i) => encoder.encode_field(&i),
+        Datum::Boolean(b) => encoder.encode_field(&b),
+        Datum::Date(v) => encoder.encode_field(&v),
+        Datum::Time(v) => encoder.encode_field(&v),
+        // FIXME: PostgreSQL does not support unsigned integers in the wire protocol.
```
Comment (jiacai2050) [inline server/src/postgresql/handler.rs:183]: Good 👍

File: server/src/postgresql/handler.rs (hunk 1 above)
Comment (holicc) [reply server/src/postgresql/handler.rs:50]: I'll change this to `PgWireError::ApiError(Box::new(e))`

File: server/src/postgresql/error.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    CreateContext { source: proxy::context::Error },
+
+    #[snafu(display("Failed to handle sql:{}, err:{}", sql, source))]
+    HandleSql { sql: String, source: GenericError },
+
+    #[snafu(display("Unexpected error, err:{}", source))]
+    Unexpected { source: std::io::Error },
+}
+
+impl From<std::io::Error> for Error {
```
[the remark this answers is not in the document]
Comment (holicc) [reply server/src/postgresql/error.rs:53]: Oops, I forgot to delete that.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Comment by jiacai2050** [comment]
Thanks, Welcome to CeresDB community 🚀

**Comment by sunng87** [comment]
Nice to see see pgwire inclusion in ceresdb. Please kindly let me know if you find anything need to change/improve.

**Comment by jiacai2050** [comment]
Thanks for your hard work on pgwire. 👍

**Pull request merged**